src/otk/transform.py

```diff
@@ -140,32 +138,27 @@ def process_defines(ctx: Context, state: State, tree: Any):
             continue
 
         if key.startswith("otk.include"):
-            # TODO: disallow this (see https://github.com/osbuild/otk/issues/116)
+            raise ParseError(f"otk.include not allowed in an otk.define in {state.path}")
+
+        if key.startswith("otk.op"):
             del tree[key]
-            # Pass {"otk.include...": path} to resolve() to have it processed recursively.
-            # The contents will become the define block.
-            incl = resolve(ctx, state, {key: value})
-            subblock.update(incl)
+            value = op(ctx, value, key)
+            ctx.define(".".join(state.define_subkeys), value)
+            continue
 
+        define_key = ".".join(state.define_subkeys+[key])
         if isinstance(value, dict):
-            # the value is a dict: process it recursively under a new subblock
-            new_subblock = {}
-            if isinstance(subblock.get(key), dict):
-                # If the new subblock already exists and is a dictionary, use it so they get merged
-                # Any other type will be replaced completely.
-                new_subblock = subblock[key]
-
-            # set the new subblock on the parent so that both context and subblock are available immediately
-            subblock[key] = new_subblock
-            new_state = state.copy(defines=new_subblock)
+            new_define_subkeys = state.define_subkeys.copy()
+            new_define_subkeys.append(key)
+            new_state = state.copy(define_subkeys=new_define_subkeys)
```

Comment:
I think this would become much easier to read if `State` gave us some helper methods for this.  I'm thinking maybe `state.define_subkeys` could return the `"a.b"` form itself and also the `copy()` method could instead do:
```python
def copy(..., subkey_add: str = None):
  new_subkeys = self._subkeys.copy()
  if subkey_add:
    new_subkeys.append(subkey_add)
  return State(..., new_subkeys)
```

and
```python
@property
def define_subkeys(self):
  return ".".join(self._subkeys)
```

Reply:
Thanks for this idea! I reworked the state encapsulation a bit, I hope it captures your suggestion.